### ISSUE TITLE
PEBAK / pacman -Syu tale of woe on XP

### ISSUE THREAD

**Issue by PhilipOakley** [git-for-windows/git#415]
This is just to log my tale of woe in case its more than a user PEBAK failure. I'll close it after a full day (i.e. Tuesday morning 22nd-Sept).

tldr; My XP system borked (corrupted my mail reader OE6 and other stuff) during a SDK update using `pacman -Syu`. I was tired, it was late. I should have waited. I'm now rebuilding the netbook. Don't know if it had anything to do with the SDK update :-( 

I was trying to update my 2.5.1 SDK install on my XPsp3 netbook using the style of https://github.com/git-for-windows/git/issues/364#issuecomment-139565017 (i.e. call pacman -Syu in your SDK to update all packages to the current version). 

Sourceforge was on a go slow, and I repeated the command a couple of times to eventually get the whole package. During that time my OE6 was open (one uses what one has..) and downloaded a few more Git List emails. However when I tried opening the emails, the OE viewer showed the mails as being blank. Worse still, all the emails in the databases showed as blank. Then Chrome started crashing (error dialog with 'close' as the only option)

Other issues may also have happend during the attempted update (recollection is hazy).

I'm now rebuilding the pc after a lot of fun collecting all the relevant content from the drives and doing a clean re-install (built-in recovery) (4 days so far). At least the old cruft will be gone...

If others have had a very similar issue then maybe there might be something to look at. Otherwise it's just 'experience

Aside: Given that sometimes the SDK updates involve all the base components, is there a generic "right way" of updating the SDK that covers all contingencies? E.g. can one always just run the https://github.com/git-for-windows/build-extra/releases/latest on top of an old install and expect sucess?

**Comment by PhilipOakley** [comment]
Follow up to the aside. I have created a Wiki page on Updating Your SDK https://github.com/git-for-windows/git/wiki/Updating-your-SDK, which is based directly on dscho's text.

The one issue there is how to know if the update requires the forced .dll updates?

**Comment by dscho** [comment]
I cannot really see how the SDK update can affect Outlook or Chrome...

**Comment by PhilipOakley** [comment]
Hi dscho, 

If I had to guess (now) at a plausible option as to how an SDK update could be an issue, it would be the little nuance hidden in the [Updating Your SDK](https://github.com/git-for-windows/git/wiki/Updating-your-SDK) where you said to beware if MSYS2 was being updated. 

It's plausible that it was being updated, but it's still, in a sense, my fault for being unprepared.

I've just about got the m/c back, so it's a learning exercise...

Is there a way, within the `pacman` command method, of determining if a particular package requires update *before* the update is initiated?

**Comment by dscho** [comment]
The warning was more about unexpected *MSys2* errors when msys2-runtime or bash are updated (heap errors)... Pacman updates should not be able to affect *anything* outside MSys2... There is simply no code in Pacman nor msys2-runtime to cause that ('cept for bugs, of course).

**Comment by PhilipOakley** [comment]
Aside: another plausible option [for the XP/OE6 crash] is that I'd reached some limit on number of emails or size of one of the email directories. Etc. These things happen. I just though it worth noting the association [correlation without causation].

In the meanwhile I've been tweaking the wiki page I created - It wasn't immediately obvious that 'Package management' had much to do with updating the broader SDK (a bit of context blindness;-)

Still to resolve how to pre-detect the possibility of an unexpected *MSys2* error. Maybe the `pacman -T` option is needed but the man page didn't provide enough illumination to see.

**Comment by dscho** [comment]
I will work on scripts to automate the individual steps necessary to deploy a new Git for Windows this week. One part will make updating the SDK safer.

**Comment by PhilipOakley** [comment]
Just a thought... Maybe don't automate the SDK update too soon, rather make sure that the steps are well described so that myself and others can follow along.

At times we (others) are at the wrong end of the USAUAOIT [unskilled and unaware of it](http://psych.colorado.edu/~vanboven/teaching/p7536_heurbias/p7536_readings/kruger_dunning.pdf) scales (i.e. $"Burden of Expertise" p1131 vs $"1st/2nd Quartiles" fig2).

There's a lot more I need to learn in some of the areas. (Isn't 'good' documentation tedious - been there, failed often ;-)

**Comment by dscho** [comment]
Well, I do not really think there is a value in every contributor having to repeat the exact same steps manually as the next one. I really would like to automate things via scripts. And of course those scripts serve as documentation as well (in particular with the comments I try to provide, e.g. in [SDK's setup script](https://github.com/git-for-windows/build-extra/blob/master/sdk-installer/setup-git-sdk.bat).

**Comment by PhilipOakley** [comment]
Sorry if I was unclear. I was more talking about finding a way of informing the contributors (so they knew where to read, etc.) rather than forcing fine grained error prone manual steps on them.

I'll see if I can improve the wiki page I started (which was based on your stirling work ;-)

**Comment by nalla** [comment]
OT: I added the call to `update-core` to https://github.com/git-for-windows/git/wiki/Updating-your-SDK to enhance the update experience. The call to update core checks if any packages that requires a restart of the _Msys2_ console have to be updated.

**Comment by PhilipOakley** [comment]
>OT: I added the call to update-core to https://github.com/git-for-windows/git/wiki/Updating-your-SDK to enhance the update experience. >The call to update core checks if any packages that requires a restart of the Msys2 console have to be updated.
That’s a useful extra, though we probably ought to also say **HOW** it is to be restarted, not just “you *MUST* restart MSys2”.
I think that restart is the side comment to close the window rather than exit that listed in the Caution section.
I was also going to add a link to the SDK's setup script <https://github.com/git-for-windows/build-extra/blob/master/sdk-installer/setup-git-sdk.bat>  that dscho had given.

**Comment by dscho** [comment]
re: `update-core` :+1:

**Comment by PhilipOakley** [comment]
I've added a link for the `setup-git-sdk` and the `core-update` into the 'updating your SDK' page. 

I see @nalla has also improved the page, though I wonder about if we should retain the original comments about what windows will be MSys2 and how to close them? I'm not sure how the 'don't use `exit`' applies, hence my asking.

**Comment by dscho** [comment]
> I'm not sure how the 'don't use exit' applies

I'm pretty certain that this rule applies because usually an `exit` after msys2-runtime was updated fails miserably, with an endless loop of error messages purporting that there is something wrong with some kind of heap.

**Comment by PhilipOakley** [comment]
It's time to close the primary 'issue'  of the "tale of woe on XP during SDK update" (as promised) as no one else has reported having any OS/programme corruption during an SDK update, but at least it will be on file if needed.

I'll re-add the tweak to the wiki page about how to close the MSys2 windows, and probably note it to upstream for the extra clarification.

Thanks for the support.

**Comment by PhilipOakley** [comment]
Just to say that I raised a PR https://github.com/Alexpux/MSYS2-pacman/pull/18 with Alexpux for the 'Prompt forst' and 'How to Close' issue above https://github.com/git-for-windows/git/issues/415#issuecomment-142016314 whicjh has now been merged upstream.

@dscho How should the G4W pacman be updated? Will you pick it up or would you like a matching PR here?

**Comment by dscho** [comment]
> @dscho How should the G4W pacman be updated? Will you pick it up or would you like a matching PR here?

It will be updated automatically because Git for Windows uses MSys2's repository for almost all packages. There are only two packages we override: `msys2-runtime` and `mingw-w64-git`. Both are slated to be reconciled eventually, pending https://github.com/git-for-windows/git/issues/284.

So: no need to do anything. The next `pacman -Syu` will pick up the newest package versions.
